slatedb/src/wal_buffer.rs

```diff
@@ -0,0 +1,602 @@
+/// current WAL) to remote storage.
+///
+/// The manager is thread-safe and can be safely shared across multiple threads.
+///
+/// Please note:
+///
+/// - the size limit (`max_wal_size`) is a soft threshold. WAL entries within a single write batch are
+///   guaranteed to be written atomically to the same WAL file.
+/// - Fatal errors during flush operations are stored internally and propagated to all subsequent
+///   operations. The manager becomes unusable after encountering a fatal error.
+pub struct WalBufferManager {
+    inner: Arc<parking_lot::RwLock<WalBufferManagerInner>>,
+    wal_id_incrementor: Arc<dyn WalIdStore + Send + Sync>,
+    quit_once: WatchableOnceCell<Result<(), SlateDBError>>,
+    table_store: Arc<TableStore>,
+    mono_clock: Arc<MonotonicClock>,
+    max_wal_bytes_size: usize,
+    max_flush_interval: Option<Duration>,
+}
+
+struct WalBufferManagerInner {
+    current_wal: Arc<KVTable>,
+    /// When the current WAL is ready to be flushed, it'll be moved to the `immutable_wals`.
+    /// The flusher will try flush all the immutable wals to remote storage.
+    immutable_wals: VecDeque<(u64, Arc<KVTable>)>,
+    /// The channel to send the flush work to the background worker.
+    flush_tx: Option<mpsc::Sender<WalFlushWork>>,
+    /// task handle of the background worker.
+    background_task: Option<JoinHandle<Result<(), SlateDBError>>>,
+    /// Whenever a WAL is applied to Memtable and successfully flushed to remote storage,
+    /// the immutable wal can be recycled in memory.
+    last_applied_seq: Option<u64>,
+    /// The flusher will update the recent_flushed_wal_id and last_flushed_seq when the flush is done.
+    recent_flushed_wal_id: u64,
+    /// The last flushed sequence number.
+    last_remote_persisted_seq: Arc<MonotonicSeq>,
+}
+
+impl WalBufferManager {
+    pub fn new(
+        wal_id_incrementor: Arc<dyn WalIdStore + Send + Sync>,
+        recent_flushed_wal_id: u64,
+        last_remote_persisted_seq: Arc<MonotonicSeq>,
+        table_store: Arc<TableStore>,
+        mono_clock: Arc<MonotonicClock>,
+        max_wal_bytes_size: usize,
+        max_flush_interval: Option<Duration>,
+    ) -> Self {
+        let current_wal = Arc::new(KVTable::new());
+        let immutable_wals = VecDeque::new();
+        let inner = WalBufferManagerInner {
+            current_wal,
+            immutable_wals,
+            last_applied_seq: None,
+            last_remote_persisted_seq,
+            recent_flushed_wal_id,
+            flush_tx: None,
+            background_task: None,
+        };
+        Self {
+            inner: Arc::new(parking_lot::RwLock::new(inner)),
+            wal_id_incrementor,
+            quit_once: WatchableOnceCell::new(),
+            table_store,
+            mono_clock,
+            max_wal_bytes_size,
+            max_flush_interval,
+        }
+    }
+
+    pub async fn start_background(self: &Arc<Self>) {
+        let (flush_tx, flush_rx) = mpsc::channel(1);
```

Comment:
Suggest doing an `if inner.background_task.is_some()` then raise error already started

Reply:
fixed in deda3a8